packages/context/src/handler-chain.ts

```diff
@@ -0,0 +1,210 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as debugFactory from 'debug';
+import {BindingFilter} from './binding-filter';
+import {BindingAddress} from './binding-key';
+import {BindingComparator} from './binding-sorter';
+import {Context} from './context';
+import {transformValueOrPromise, ValueOrPromise} from './value-promise';
+
+const debug = debugFactory('loopback:context:handler-chain');
+
+/**
+ * Return value for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationResult = any;
+
+/**
+ * Array of arguments for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationArgs = any[];
+
+/**
+ * `next` function
+ */
+export type Next = () => ValueOrPromise<InvocationResult>;
+
+/**
+ * A handler function to be invoked in a chain for the given context
+ */
+export interface Handler<C extends Context = Context> {
```

Comment:
I find the name `Handler` rather ambiguous. Also in my mind, handler is responsible for handling the entire process, not calling `next`. I.e. when I see the term `Handler`, I expect something closer to Express route handlers (functions call `res.send` and don't call `next`). In your proposal, Handler is closer to Express middleware (functions that manipulate context and call `next`).

Can we come up with a better name? Have you considered using the term `Middleware`? It's the term that Koa is using for this kind of functions. See https://www.npmjs.com/package/koa#middleware

Reply:
I'm concerned that our users will mistakenly assume it's Express/Koa middleware.

Reply:
`Handler` is used in https://en.wikipedia.org/wiki/Chain-of-responsibility_pattern.

Reply:
> I'm concerned that our users will mistakenly assume it's Express/Koa middleware.

Well, the interface does describe Koa middleware too, doesn't it?

Would `GenericMiddleware<T>` alleviate your concern?